application/preprocessing/generate_features.py

```diff
@@ -0,0 +1,190 @@
+def __get_word_count(lyrics):
+    total = 0
+    for word in lyrics:
+        total += word[1]
+    return total
+
+# helper to create percentage histogram
+def __get_frequency_histogram(values):
+    hist = numpy.histogram(values, bins=numpy.linspace(0, 1, 11))
+    return {
+        'frequencies': hist[0].tolist(),
+        'bin_edges': hist[1].tolist()
+    }
+
+# calculates the parts of speech the words in a song are (as percentage of total words)
+# for the song, it saves them like the following:
+# parts_of_speech = {
+#    noun: .15,
+#    verb: .1
+# }
+# for the genre it saves the histogram of song distributions like so:
+# noun_hist = {
+#     frequencies: {
+#         .21,
+#         ...
+#         .01
+#     }
+#     bin_edges: {
+#         .0
+#         ...
+#         1.0
+#     }
+def calculate_word_tags(frame, genre_data, song_data):
+    # limit columns to the ones we care about
+    frame = frame[['track_id', 'lyrics']]
+
+    # list of wanted tag types (missing punctuation as that should not be in the dataset it seems)
+    tag_list = ['adj', 'adp', 'adv', 'conj', 'det', 'noun', 'num', 'prt', 'pron', 'verb', 'x']
+    genre_parts = dict(zip(tag_list, [[] for _ in range(len(tag_list))]))
+
+    for index, song in frame.iterrows():
+        song_name = song['track_id']
+
+        lyrics = __get_lyrics(song)
+
+        speech_counter = Counter()
+        for pair in lyrics:
+            part_of_speech = nltk_helper.get_word_tag(top_words.unstemmed[pair[0]])
+            if(part_of_speech):
+                speech_counter[part_of_speech.lower()] += pair[1]
+
+        counts = [speech_counter[x] for x in tag_list]
+        frequencies = [x/__get_word_count(lyrics) for x in counts]
+
+        for tag, frequency in zip(tag_list, frequencies):
+            genre_parts[tag].append(frequency)
+
+        song_data[song_name]['parts_of_speech'] = dict(zip(tag_list, frequencies))
+
+    for tag in tag_list:
+        genre_data[tag.lower() + '_hist'] = __get_frequency_histogram(genre_parts[tag])
+
+
+# save durations of songs in the songs and a histogram in the genre
+def calculate_duration(frame, genre_data, song_data):
+    # limit columns to the ones we care about
+    frame = frame[['track_id', 'duration']]
+    durations = list(frame['duration'])
+
+    hist = numpy.histogram(durations)
+
+    genre_data['duration_hist'] = {
+        'frequencies': hist[0].tolist(),
+        'bin_edges': hist[1].tolist()
+    }
+
+    for index, song in frame.iterrows():
+        song_name = song['track_id']
+        song_data[song_name]['duration'] = song['duration']
+
+
+# calculates the top words for a genre and saves them to the genre
+def calculate_popular_words(frame, genre_data, song_data):
+    word_popularity = Counter()
+    song_count = 0
+    for index, song in frame.iterrows():
+        song_count += 1
+        lyrics = __get_lyrics(song)
+        for word in lyrics:
+            word_popularity[word[0]] += word[1]
```

Comment:
Depending on how the data is saved in `lyrics`, make sure you don't fall into the same trap I did where I thought word and frequency were reversed.

Reply:
If you've tested this on real data and it looks good then I wouldn't worry about it, I'm just mentioning it because I ran into the same problem literally 5 minutes ago.